source/v5/mqtt3_to_mqtt5_adapter.c

```diff
@@ -1651,7 +1681,549 @@ static uint16_t s_aws_mqtt_client_connection_5_publish(
+
+    struct aws_mqtt_topic_subscription subscription = {
+        .topic = *topic_filter,
+        .qos = qos,
+        .on_publish = on_publish,
+        .on_cleanup = on_ud_cleanup,
+        .on_publish_ud = on_publish_ud,
+    };
+
+    struct aws_mqtt3_to_mqtt5_adapter_subscribe_options subscribe_options = {
+        .adapter = adapter,
+        .subscriptions = &subscription,
+        .subscription_count = 1,
+        .on_suback = on_suback,
+        .on_suback_user_data = on_suback_user_data,
+    };
+
+    struct aws_mqtt3_to_mqtt5_adapter_operation_subscribe *operation =
+        aws_mqtt3_to_mqtt5_adapter_operation_new_subscribe(adapter->allocator, &subscribe_options);
+    if (operation == NULL) {
+        return 0;
+    }
+
+    if (aws_mqtt3_to_mqtt5_adapter_operation_table_add_operation(&adapter->operational_state, &operation->base)) {
+        goto error;
+    }
+
+    uint16_t synthetic_id = operation->base.id;
+
+    /* add a transient reference to the adapter until this operation is passed into the event loop */
+    aws_mqtt3_to_mqtt5_adapter_operation_reference_adapter(&operation->base);
+
+    /* add a transient reference to the operation until this operation is passed into the event loop */
+    aws_mqtt3_to_mqtt5_adapter_operation_acquire(&operation->base);
+
+    aws_task_init(
+        &operation->base.submission_task,
+        s_adapter_subscribe_submission_fn,
+        operation,
+        "Mqtt3ToMqtt5AdapterSubscribeSubmission");
+
+    aws_event_loop_schedule_task_now(adapter->loop, &operation->base.submission_task);
+
+    return synthetic_id;
+
+error:
+
+    aws_mqtt3_to_mqtt5_adapter_operation_release(&operation->base);
+
+    return 0;
+}
+
+static uint16_t s_aws_mqtt_client_connection_5_subscribe_multiple(
+    void *impl,
+    const struct aws_array_list *topic_filters,
+    aws_mqtt_suback_multi_fn *on_suback,
+    void *on_suback_user_data) {
+
+    struct aws_mqtt_client_connection_5_impl *adapter = impl;
+
+    struct aws_mqtt_topic_subscription *subscriptions = topic_filters->data;
+
+    struct aws_mqtt3_to_mqtt5_adapter_subscribe_options subscribe_options = {
+        .adapter = adapter,
+        .subscriptions = subscriptions,
+        .subscription_count = aws_array_list_length(topic_filters),
+        .on_multi_suback = on_suback,
+        .on_multi_suback_user_data = on_suback_user_data,
+    };
+
+    struct aws_mqtt3_to_mqtt5_adapter_operation_subscribe *operation =
+        aws_mqtt3_to_mqtt5_adapter_operation_new_subscribe(adapter->allocator, &subscribe_options);
+    if (operation == NULL) {
+        return 0;
+    }
+
+    if (aws_mqtt3_to_mqtt5_adapter_operation_table_add_operation(&adapter->operational_state, &operation->base)) {
+        goto error;
+    }
+
+    uint16_t synthetic_id = operation->base.id;
+
+    /* add a transient reference to the adapter until this operation is passed into the event loop */
+    aws_mqtt3_to_mqtt5_adapter_operation_reference_adapter(&operation->base);
```

Comment:
Trivial: Apologize for being so hung up on this but it's like a splinter in my brain now. `aws_mqtt3_to_mqtt5_adapter_operation_reference_adapter` and `aws_mqtt3_to_mqtt5_adapter_operation_acquire` are always called one after the other and both work off the same base. Would it make any sense at all to handle both releases in a singular function that handles the transient references? Just a thought since this is a repeating pattern in 4 places.

Reply:
Updated to use a cross_thread_refs acquire and release function pair